### PR TITLE
improvements to regression framework

### PR DESCRIPTION
**Pull request by jbloom22**
- removed relics of command line
- improved python options, e.g. covarites given as list of string
- factored out common ops like covariate matrix code to stats/RegressionUtils
- converted LogReg to mapAnnotations, consistent with linreg and lmmreg
- removed LinearRegressionCommand and LogisticRegressionCommand

Question: here I pass the covariates from Python to Scala as an Array of String, then merge the string on comma for the parser.  We could pass it as a List of String, making jlist like jarray.  In either case, I think our parsers should take the sequence, rather than have to re-split. Thoughts?

### CODE REVIEW

**Comment by jbloom22** [comment]
I've tested that the code still works from within iPython by hand, but we'll need to automate Python tests too. I'm not sure if we've done that elsewhere yet.

**Comment by jbloom22** [comment]
Worked out the kinks in Python. Here's a working example:
```
import pyhail
hc = pyhail.HailContext()
vds = (hc.import_vcf('src/test/resources/regressionLogistic.vcf')
    .annotate_samples_table('src/test/resources/regressionLogistic.cov', 'Sample', root='sa.cov', config=pyhail.TextTableConfig(impute=True))
    .annotate_samples_table('src/test/resources/regressionLogisticBoolean.pheno', 'Sample', root='sa.pheno', config=pyhail.TextTableConfig(types='isCase: Boolean', missing='0'))
    .logreg('wald', 'sa.pheno.isCase', ['sa.cov.Cov1', 'sa.cov.Cov2'])
    .linreg('sa.pheno.isCase', ['sa.cov.Cov1', 'sa.cov.Cov2']))
```

**Comment by jbloom22** [comment]
This should go in before the `jb_lmm_getthisin` branch, which is rebased off this one, continuing the unification of lin, log, and mixed regression, with more refactoring.

(alternatively, the `jb_lmm_getthisin` branch supercedes this one)

**Comment by jbloom22** [comment]
I've rebased to move from pyhail to hail.

**Comment by jbloom22** [comment]
Rebased to is.hail

File: python/hail/dataset.py [hunk 1]
```diff
@@ -2039,28 +2039,26 @@ def linreg(self, y, covariates='', root='va.linreg', minac=1, minaf=None):
 
         :param str y: Response sample annotation.
 
-        :param str covariates: Covariant sample annotations, comma separated.
+        :param covariates: list of covariant sample annotations
```
Comment (cseed) [inline python/hail/dataset.py]: These aren't just annotations, these are expressions, right?  You use this language in several places.

File: src/main/scala/is/hail/methods/LogisticRegression.scala [hunk 2]
```diff
@@ -7,29 +7,36 @@ import is.hail.annotations.Annotation
...
-      fatal(s"$n samples and $k ${plural(k, "covariate")} with intercept implies $d degrees of freedom.")
+    val logRegTest = tests(test)
+
+    val (y, cov, sampleMask) = getPhenoCovCompleteSamples(vds, ySA, covSA)
+
+    if (! y.forall(yi => yi == 0d || yi == 1d))
+      fatal(s"For logistic regression, phenotype must be Boolean or numeric with all values equal to 0 or 1")
+
+    if (!tests.isDefinedAt(test))
+      fatal(s"Supported tests are ${tests.keys.mkString(", ")}, got: $test")
```
Comment (cseed) [inline src/main/scala/is/hail/methods/LogisticRegression.scala]: Spaces around expression in `${ }`.  Will be inserted by auto-indent.

File: src/main/scala/is/hail/stats/RegressionUtils.scala [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+          rows = n,
+          cols = 1 + k,
+          data = covArray,
+          offset = 0,
+          majorStride = 1 + k,
+          isTranspose = true)
+
+//    val r = rank(cov)
+//    if (r < k + 1)
+//      fatal(s"Covariates and intercept are not linearly independent: total rank is $r (less than 1 + $k covariates)")
```
Comment (cseed) [inline src/main/scala/is/hail/stats/RegressionUtils.scala]: Resolve.

**Comment by cseed** [comment]
Taking a string or list of string as covariates and passing a string after doing `','.join(covariates)` on the python side is what we do for the annotation commands.  This seems as good as anything.

**Comment by jbloom22** [comment]
I made the requested changes, improved the docs a bit more, removed the rank testing for now (may bring back in later PR), and am keeping the covariate field as list of expressions, even for one expression. I feel this is more natural in the regression setting, where one can even have no covariates (default is []), and I prefer to pass an array of string into RegressionUtils, though we should make another pass on parsing later (right now, I do still concatenate with ',' just before sending into the parser but would to not merge only to split).

**Pull request merged**